### PR TITLE
Allow custom age groups sizes in `create_age_groups`

### LINKED ISSUE #93: `create_age_groups` with bins of custom lengths
**Is your feature request related to a problem? Please describe.**
At my team, we have to create age bands which do not have the same length. Currently, the `create_age_groups` does not allow this.

**Describe the solution you'd like**
I created a simple version of `create_age_groups` which allows custom bin lengths by specifying the cut points with a numeric vector.

### LINKED ISSUE #135: Improve `create_age_groups`
Hi, I think we can improve the speed of `create_age_groups` quite a bit and also remove the dependency on 'utils' package.

If we avoid `cut()` which is inefficient in creating factors as it goes through unnecessary `unique()` + `match()` steps internally. 
We already have our cleaned age breaks which are unique and sorted, meaning we can avoid using `cut()` and directly use `.bincode()`.
`.bincode()` is basically a low-level factor constructor and also what `cut()` uses as well.
To get a character vector, all that's needed is to subset our age breaks onto our bin codes.

On the topic of `cut()` inefficiency, there is a stack thread I opened a while ago: https://stackoverflow.com/questions/76867914/can-cut-be-improved

Proposed function and benchmark:

``` r
create_age_groups <- function(x, from = 0, to = 90, by = 5, as_factor = FALSE){
  
  if (!is.numeric(x)) {
    cli::cli_abort("{.arg x} must be a {.cls numeric} vector, not a {.cls {class(x)}} vector.")
  }
  
  breaks <- seq(from, to, by)
  breaks <- sort(unique(breaks))
  n_breaks <- length(breaks)
  n <- max(n_breaks - 1L, 0L)
  
  bands <- paste0(
    breaks[seq_len(n)], "-", 
    breaks[seq.int(to = n_breaks, length.out = n)] - 1L
  )
  
  rightmost_band <- paste0(breaks[n_breaks], "+")
  bands[n_breaks] <- rightmost_band
  
  codes <- .bincode(x, breaks = c(breaks, Inf), right = FALSE)
  
  if (as_factor) {
    out <- codes
    levels(out) <- bands
    class(out) <- c("ordered", "factor")
  }
  else {
    out <- bands[codes]
  }
  out
}
library(bench)

x <- 20
create_age_groups(x)
#> [1] "20-24"
phsmethods::create_age_groups(x)
#> [1] "20-24"

mark(create_age_groups(x), 
     phsmethods::create_age_groups(x))
#> # A tibble: 2 × 6
#>   expression                            min  median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                       <bch:tm> <bch:t>     <dbl> <bch:byt>    <dbl>
#> 1 create_age_groups(x)               98.6µs   103µs     9202.    2.41KB    10.4 
#> 2 phsmethods::create_age_groups(x)  155.5µs   168µs     5697.    3.84KB     8.30
mark(create_age_groups(x, as_factor = TRUE), 
     phsmethods::create_age_groups(x, as_factor = TRUE))
#> # A tibble: 2 × 6
#>   expression                             min median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                           <bch> <bch:>     <dbl> <bch:byt>    <dbl>
#> 1 create_age_groups(x, as_factor = TR… 101µs  106µs     8998.    2.71KB    12.7 
#> 2 phsmethods::create_age_groups(x, as… 153µs  162µs     5868.    3.84KB     8.30

x <- sample(0:100, 10^7, T)
mark(create_age_groups(x), 
     phsmethods::create_age_groups(x))
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 2 × 6
#>   expression                            min  median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                       <bch:tm> <bch:t>     <dbl> <bch:byt>    <dbl>
#> 1 create_age_groups(x)                414ms   416ms      2.40     191MB     2.40
#> 2 phsmethods::create_age_groups(x)    701ms   701ms      1.43     420MB     4.28
mark(create_age_groups(x, as_factor = TRUE), 
     phsmethods::create_age_groups(x, as_factor = TRUE))
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 2 × 6
#>   expression                             min median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                           <bch> <bch:>     <dbl> <bch:byt>    <dbl>
#> 1 create_age_groups(x, as_factor = TR… 361ms  361ms      2.77     153MB     2.77
#> 2 phsmethods::create_age_groups(x, as… 654ms  654ms      1.53     343MB     1.53
```

<sup>Created on 2024-09-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

This obviously relates to issues #93 and #54, which I think are also worthwhile but as subsequent step.

### PR DESCRIPTION
**Pull request by Moohan**
This alters the interface of `create_age_groups` to both allow non-uniform age groups and to simplify it, going from 5 parameters to 3.

I implemented it as discussed in #93.

I've also implemented @Nic-Chr suggestions from #135, which makes the whole process more efficient - slightly faster and reasonably more memory efficient.

The 'old' params will still work but issue a warning, the `from`, `to` and `by` are translated into the new `breaks` param as `seq(from, to, by)`. We can deprecate further to an informative error in a future version, and finally remove them, I think a year between each phase would be fine.

I added some input validation for the new and existing params, to error or warn on invalid input; wrong types, negative ages, or possibly ambiguous things like fractional ages.

Finally I added tests for the new behaviour and the deprecation warnings.

### CODE REVIEW

File: R/create_age_groups.R [hunk 1]
```diff
@@ -5,47 +5,71 @@
...
+  from = lifecycle::deprecated(),
+  to = lifecycle::deprecated(),
+  by = lifecycle::deprecated(),
+  breaks = seq(0, 90, 5),
+  as_factor = FALSE
+) {
+  if (
+    lifecycle::is_present(from) ||
+      lifecycle::is_present(to) ||
+      lifecycle::is_present(by)
```
Comment (Moohan) [inline R/create_age_groups.R:111]: Use `!missing(...)` instead of `lifecycle::is_present` (check this is the same).

Comment (Nic-Chr) [reply R/create_age_groups.R:111]: I checked and the overhead of `lifecycle::is_present` is very small anyway so we can keep it I think.

**Comment by Nic-Chr** [comment]
As mentioned in #135 we can safely use `.bincode()` instead of `cut()` to improve performance.

``` r
create_age_groups1 <- function(
    x,
    from = lifecycle::deprecated(),
    to = lifecycle::deprecated(),
    by = lifecycle::deprecated(),
    as_factor = FALSE,
    breaks = seq(0, 90, 5)
) {
  if (
    lifecycle::is_present(from) ||
    lifecycle::is_present(to) ||
    lifecycle::is_present(by)
  ) {
    if (missing(breaks)) {
      lifecycle::deprecate_soft(
        when = "1.0.0",
        what = I("create_age_groups(from/to/by)"),
        with = "create_age_groups(breaks)"
      )
      
      # Fall back to defaults for any missing deprecated arguments
      if (!lifecycle::is_present(from)) from <- 0
      if (!lifecycle::is_present(to)) to <- 90
      if (!lifecycle::is_present(by)) by <- 5
      
      breaks <- seq(from, to, by)
    } else {
      cli::cli_abort(
        "{.arg from}, {.arg to} or {.arg by} should no longer be used, use {.arg breaks} only"
      )
    }
  }
  
  breaks <- c(breaks, Inf)
  breaks <- sort(unique(breaks))
  
  # Create labels based on consecutive values in breaks
  labels <- paste0(utils::head(breaks, -1), "-", utils::tail(breaks, -1) - 1)
  
  # Reformat label for last value
  labels <- gsub("-Inf", "+", labels)
  
  agegroup <- cut(
    x,
    breaks = breaks,
    labels = labels,
    right = FALSE,
    ordered_result = TRUE
  )
  
  if (as_factor == FALSE) {
    agegroup <- as.character(agegroup)
  }
  
  agegroup
}
create_age_groups2 <- function(
    x,
    from = lifecycle::deprecated(),
    to = lifecycle::deprecated(),
    by = lifecycle::deprecated(),
    as_factor = FALSE,
    breaks = seq(0, 90, 5)
) {
  if (
    lifecycle::is_present(from) ||
    lifecycle::is_present(to) ||
    lifecycle::is_present(by)
  ) {
    if (missing(breaks)) {
      lifecycle::deprecate_soft(
        when = "1.0.0",
        what = I("create_age_groups(from/to/by)"),
        with = "create_age_groups(breaks)"
      )
      
      # Fall back to defaults for any missing deprecated arguments
      if (!lifecycle::is_present(from)) from <- 0
      if (!lifecycle::is_present(to)) to <- 90
      if (!lifecycle::is_present(by)) by <- 5
      
      breaks <- seq(from, to, by)
    } else {
      cli::cli_abort(
        "{.arg from}, {.arg to} or {.arg by} should no longer be used, use {.arg breaks} only"
      )
    }
  }
  
  
  n_breaks <- length(breaks)
  n_intervals <- max(n_breaks - 1L, 0L)
  
  age_bands <- paste0(
    breaks[seq_len(n_intervals)], "-",
    breaks[seq(to = n_breaks, length.out = n_intervals)] - 1L
  )
  rightmost_band <- paste0(breaks[n_breaks], "+")
  age_bands[n_breaks] <- rightmost_band
  
  factor_codes <- .bincode(x, breaks = c(breaks, Inf), right = FALSE)
  
  if (as_factor){
    out <- factor_codes
    levels(out) <- age_bands
    class(out) <- c("ordered", "factor")
  } else {
    out <- age_bands[factor_codes]
  }
  out
}

x <- sample(0:100, 10^6, TRUE)

library(bench)
mark(
  with_cut = create_age_groups1(x, as_factor = FALSE),
  with_bincode = create_age_groups2(x, as_factor = FALSE)
)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 2 × 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 with_cut       46.1ms     48ms      19.9    30.8MB     21.9
#> 2 with_bincode   37.1ms   42.4ms      21.1    19.3MB     17.3

mark(
  with_cut = create_age_groups1(x, as_factor = TRUE),
  with_bincode = create_age_groups2(x, as_factor = TRUE)
)
#> # A tibble: 2 × 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 with_cut       36.6ms   36.6ms      27.3    22.9MB    150. 
#> 2 with_bincode   32.1ms   32.3ms      30.3    15.3MB     45.4

# Overhead on small vectors
y <- 1:10

mark(
  with_cut = create_age_groups1(y, as_factor = FALSE),
  with_bincode = create_age_groups2(y, as_factor = FALSE)
)
#> # A tibble: 2 × 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 with_cut      162.8µs  169.3µs     5719.    3.94KB    10.3 
#> 2 with_bincode   78.8µs   82.1µs    11370.    1.73KB     8.22

mark(
  with_cut = create_age_groups1(y, as_factor = TRUE),
  with_bincode = create_age_groups2(y, as_factor = TRUE)
)
#> # A tibble: 2 × 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 with_cut      158.9µs  167.1µs     5517.    3.94KB     10.4
#> 2 with_bincode   80.5µs   83.8µs    11519.    2.02KB     10.4
```

<sup>Created on 2026-02-26 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Note:** I replaced the use of `utils::tail` and `utils::head` to improve overhead.

The benchmarks show slight improvement in speed, fair improvement on memory usage. Given how easy of an implementation it is with essentially no drawbacks, it might be worth implementing.

**Review by Tina815 — COMMENTED** [review]
Sorry taking a while. The function looks all good to me!

File: R/create_age_groups.R [hunk 2]
```diff
@@ -5,66 +5,146 @@
...
+      "{.arg x} must be an {.cls integer} vector, not a {.cls {class(x)}} vector."
+    )
+  }
+  if (any(x < 0, na.rm = TRUE)) {
+    cli::cli_abort("{.arg x} cannot contain negative ages.")
+  }
+  if (!is.integer(x)) {
+    x_int <- as.integer(x)
+
+    if (any(x != x_int, na.rm = TRUE)) {
```
Comment (Nic-Chr) [inline R/create_age_groups.R]: This is too strict, for example this fails: `create_age_groups(sqrt(2)^2)`
Would suggest maybe using a tolerance
`any(abs(x - x_int) < sqrt(.Machine$double.eps))`

Comment (Nic-Chr) [reply R/create_age_groups.R]: The canonical way to do it would be by rounding, because coercing to integer will truncate. 

```r
> as.integer(2 - sqrt(.Machine$double.eps))
[1] 1
```

So would suggest doing `any(abs(x - round(x)) < sqrt(.Machine$double.eps))`

Comment (Nic-Chr) [reply R/create_age_groups.R]: Other than that all looks good to me!

Comment (Tina815) [reply R/create_age_groups.R]: > This is too strict, for example this fails: `create_age_groups(sqrt(2)^2)` Would suggest maybe using a tolerance `any(abs(x - x_int) < sqrt(.Machine$double.eps))`

I received the warning for this, but it still works okay.

Comment (Nic-Chr) [reply R/create_age_groups.R]: Mistook the warning for an error, but I do think there should probably be an approximate check if we're gating on whole numbers.

File: R/create_age_groups.R [hunk 3]
```diff
@@ -5,66 +5,135 @@
...
+  by = lifecycle::deprecated(),
+  as_factor = FALSE,
+  breaks = seq(0, 90, 5)
+) {
+  if (!is.numeric(x)) {
+    cli::cli_abort(
+      "{.arg x} must be an {.cls integer} vector, not a {.cls {class(x)}} vector."
+    )
+  }
+  if (any(x < 0, na.rm = TRUE)) {
```
Comment (Nic-Chr) [inline R/create_age_groups.R]: `isTRUE(min(x, na.rm = TRUE) < 0)` would avoid the intermediate logical vector.

Comment (Tina815) [reply R/create_age_groups.R]: Can we only have `min(x, na.rm = TRUE) < 0` without isTRUE, since `na` are removed? I would agree this way is more efficient, but `any(x < 0, na.rm = TRUE)` is more readable and intuitive?

Comment (Moohan) [reply R/create_age_groups.R]: I've made that change and generally improved the checks by checking for NAs first (no NA breaks and at least one non-NA x), also checking that breaks are positive whole numbers.

Comment (Tina815) [reply R/create_age_groups.R]: > I've made that change and generally improved the checks by checking for NAs first (no NA breaks and at least one non-NA x), also checking that breaks are positive whole numbers.

Thanks! I also added some tests for x all missing values, and any missing/negative values in breaks.

**Review by Tina815 — APPROVED** [review]
